File: src/app/models/disk_monitor.py
"""Disk IO monitor entities."""

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from .base import BaseModel


class DiskIoRecord(BaseModel):
    __tablename__ = "disk_io_records"

    execution_id = Column(Integer, ForeignKey("executions.id", ondelete="SET NULL"), nullable=True, index=True)
    task_id = Column(Integer, ForeignKey("tasks.id", ondelete="SET NULL"), nullable=True, index=True)
    job_id = Column(Integer, ForeignKey("jobs.id", ondelete="SET NULL"), nullable=True, index=True)
    task_name = Column(Text, nullable=True)
    job_name = Column(Text, nullable=True)
    task_type = Column(String(20), nullable=False, default="report")
    is_manual = Column(Boolean, nullable=False, default=False)
    provider = Column(String(20), nullable=False, default="unknown")
    started_at = Column(DateTime, nullable=True)
    finished_at = Column(DateTime, nullable=True)
    duration_ms = Column(Integer, nullable=True)
    backend_read_bytes = Column(Integer, nullable=True)
    backend_write_bytes = Column(Integer, nullable=True)
    weflow_read_bytes = Column(Integer, nullable=True)
    weflow_write_bytes = Column(Integer, nullable=True)
    weflow_captured = Column(Boolean, nullable=False, default=False)
    weflow_process = Column(Text, nullable=True)
    exported_file_bytes = Column(Integer, nullable=False, default=0)
    chatlog_decrypt_write_bytes = Column(Integer, nullable=False, default=0)
    chatlog_decrypt_status = Column(String(20), nullable=True)
    chatlog_work_dir = Column(Text, nullable=True)
    disk_write_bytes = Column(Integer, nullable=False, default=0)
    total_read_bytes = Column(Integer, nullable=False, default=0)
    total_write_bytes = Column(Integer, nullable=False, default=0)
    job_alert_threshold_bytes = Column(Integer, nullable=False, default=0)
    job_alert_triggered = Column(Boolean, nullable=False, default=False)
    job_alert_sent = Column(Boolean, nullable=False, default=False)
    job_alert_error = Column(Text, nullable=True)
    media_enabled = Column(Boolean, nullable=False, default=False)
    is_warning = Column(Boolean, nullable=False, default=False)
    warning_reason = Column(Text, nullable=True)
    raw_snapshot = Column(Text, nullable=True)

    execution = relationship("Execution")


class DiskInspectionJob(BaseModel):
    __tablename__ = "disk_inspection_jobs"

    name = Column(String(120), nullable=False)
    is_enabled = Column(Boolean, nullable=False, default=True)
    schedule_type = Column(String(30), nullable=False, default="daily")
    schedule_time = Column(String(5), nullable=False, default="09:00")
    schedule_weekday = Column(Integer, nullable=True)
    schedule_month_day = Column(Integer, nullable=True)
    cron_expression = Column(String(120), nullable=True)
    interval_enabled = Column(Boolean, nullable=False, default=False)
    interval_minutes = Column(Integer, nullable=True)
    window_start = Column(String(5), nullable=False, default="00:00")
    window_end = Column(String(5), nullable=False, default="24:00")
    window_hours = Column(Integer, nullable=False, default=24)
    threshold_bytes = Column(Integer, nullable=False, default=104857600)
    webhook_ids = Column(Text, nullable=False, default="[]")
    cooldown_hours = Column(Integer, nullable=False, default=6)
    weflow_process_names = Column(Text, nullable=True)
    last_run_at = Column(DateTime, nullable=True)
    next_run_at = Column(DateTime, nullable=True)
    last_alert_at = Column(DateTime, nullable=True)


class DiskInspectionRun(BaseModel):
    __tablename__ = "disk_inspection_runs"

    inspection_job_id = Column(Integer, ForeignKey("disk_inspection_jobs.id", ondelete="SET NULL"), nullable=True, index=True)
    inspection_job_name = Column(Text, nullable=True)
    status = Column(String(20), nullable=False, default="success")
    inspected_at = Column(DateTime, nullable=False)
    window_start = Column(DateTime, nullable=False)
    window_end = Column(DateTime, nullable=False)
    window_hours = Column(Integer, nullable=False, default=24)
    threshold_bytes = Column(Integer, nullable=False, default=104857600)
    total_write_bytes = Column(Integer, nullable=False, default=0)
    total_read_bytes = Column(Integer, nullable=False, default=0)
    max_single_write_bytes = Column(Integer, nullable=False, default=0)
    warning_count = Column(Integer, nullable=False, default=0)
    record_count = Column(Integer, nullable=False, default=0)
    chatlog_decrypt_count = Column(Integer, nullable=False, default=0)
    weflow_media_count = Column(Integer, nullable=False, default=0)
    alert_sent = Column(Boolean, nullable=False, default=False)
    alert_error = Column(Text, nullable=True)
    webhook_ids = Column(Text, nullable=True)
    summary = Column(Text, nullable=True)

    inspection_job = relationship("DiskInspectionJob")

File: web/src/services/diskMonitor.ts
import apiClient from "./apiClient";
import type { ApiResponse } from "./apiClient";

export type DiskIoRecord = {
  id: number;
  created_at: string;
  updated_at: string;
  execution_id?: number | null;
  task_id?: number | null;
  job_id?: number | null;
  task_name?: string | null;
  job_name?: string | null;
  task_type: string;
  is_manual: boolean;
  provider: string;
  started_at?: string | null;
  finished_at?: string | null;
  duration_ms?: number | null;
  backend_read_bytes?: number | null;
  backend_write_bytes?: number | null;
  weflow_read_bytes?: number | null;
  weflow_write_bytes?: number | null;
  weflow_captured: boolean;
  weflow_process?: string | null;
  exported_file_bytes: number;
  chatlog_decrypt_write_bytes: number;
  chatlog_decrypt_status?: string | null;
  chatlog_work_dir?: string | null;
  disk_write_bytes: number;
  total_read_bytes: number;
  total_write_bytes: number;
  media_enabled: boolean;
  is_warning: boolean;
  warning_reason?: string | null;
  job_alert_threshold_bytes: number;
  job_alert_triggered: boolean;
  job_alert_sent: boolean;
  job_alert_error?: string | null;
  raw_snapshot?: Record<string, unknown> | null;
};

export type DiskIoSummary = {
  label: string;
  hours?: number | null;
  start_time: string;
  end_time: string;
  total_write_bytes: number;
  total_read_bytes: number;
  max_single_write_bytes: number;
  warning_count: number;
  record_count: number;
  chatlog_decrypt_count: number;
  weflow_media_count: number;
};

export type DiskIoRecordPage = {
  items: DiskIoRecord[];
  total: number;
  page: number;
  page_size: number;
};

export type DiskIoQuery = {
  page: number;
  page_size: number;
  execution_id?: number;
  task_id?: number;
  job_id?: number;
  provider?: string;
  task_type?: string;
  is_warning?: boolean;
  start_time?: string;
  end_time?: string;
  sort_by?: "disk_write_bytes" | "exported_file_bytes";
  sort_order?: "asc" | "desc";
};

export type DiskIoAggregate = {
  summary: {
    disk_write_bytes: number;
    max_single_disk_write_bytes: number;
    record_count: number;
    warning_count: number;
  };
  monthly_trend: Array<{
    month: string;
    disk_write_bytes: number;
    record_count: number;
  }>;
  task_ranking: Array<{
    task_name?: string | null;
    job_name?: string | null;
    record_count: number;
    disk_write_bytes: number;
    avg_disk_write_bytes: number;
  }>;
};

export type DiskInspectionJob = {
  id: number;
  created_at: string;
  updated_at: string;
  name: string;
  is_enabled: boolean;
  schedule_type: string;
  schedule_time: string;
  schedule_weekday?: number | null;
  schedule_month_day?: number | null;
  cron_expression?: string | null;
  interval_enabled: boolean;
  interval_minutes?: number | null;
  window_start: string;
  window_end: string;
  window_hours: number;
  threshold_bytes: number;
  webhook_ids: number[];
  cooldown_hours: number;
  weflow_process_names?: string | null;
  last_run_at?: string | null;
  next_run_at?: string | null;
  last_alert_at?: string | null;
};

export type DiskInspectionJobPayload = Omit<DiskInspectionJob, "id" | "created_at" | "updated_at" | "last_run_at" | "next_run_at" | "last_alert_at">;

export type DiskInspectionRun = {
  id: number;
  created_at: string;
  updated_at: string;
  inspection_job_id?: number | null;
  inspection_job_name?: string | null;
  status: string;
  inspected_at: string;
  window_start: string;
  window_end: string;
  window_hours: number;
  threshold_bytes: number;
  total_write_bytes: number;
  total_read_bytes: number;
  max_single_write_bytes: number;
  warning_count: number;
  record_count: number;
  chatlog_decrypt_count: number;
  weflow_media_count: number;
  alert_sent: boolean;
  alert_error?: string | null;
  webhook_ids?: number[] | null;
  summary?: string | null;
};

export type DiskInspectionRunPage = {
  items: DiskInspectionRun[];
  total: number;
  page: number;
  page_size: number;
};

export type DiskInspectionRunQuery = {
  page: number;
  page_size: number;
  inspection_job_id?: number;
  status?: string;
  start_time?: string;
  end_time?: string;
};

export async function fetchDiskIoRecords(params: DiskIoQuery): Promise<DiskIoRecordPage> {
  const response = await apiClient.get<ApiResponse<DiskIoRecordPage>>("/api/disk-monitor/records", { params });
  return response.data.data;
}

export async function fetchDiskIoSummaries(): Promise<DiskIoSummary[]> {
  const response = await apiClient.get<ApiResponse<DiskIoSummary[]>>("/api/disk-monitor/records/summary");
  return response.data.data;
}

export async function fetchDiskIoAggregate(params: { start_time: string; end_time: string }): Promise<DiskIoAggregate> {
  const response = await apiClient.get<ApiResponse<DiskIoAggregate>>("/api/disk-monitor/records/aggregate", { params });
  return response.data.data;
}

export async function fetchDiskIoRecord(recordId: number): Promise<DiskIoRecord> {
  const response = await apiClient.get<ApiResponse<DiskIoRecord>>(`/api/disk-monitor/records/${recordId}`);
  return response.data.data;
}

export async function fetchDiskInspectionJobs(): Promise<DiskInspectionJob[]> {
  const response = await apiClient.get<ApiResponse<DiskInspectionJob[]>>("/api/disk-monitor/inspection-jobs");
  return response.data.data;
}

export async function createDiskInspectionJob(payload: DiskInspectionJobPayload): Promise<DiskInspectionJob> {
  const response = await apiClient.post<ApiResponse<DiskInspectionJob>>("/api/disk-monitor/inspection-jobs", payload);
  return response.data.data;
}

export async function updateDiskInspectionJob(jobId: number, payload: DiskInspectionJobPayload): Promise<DiskInspectionJob> {
  const response = await apiClient.put<ApiResponse<DiskInspectionJob>>(`/api/disk-monitor/inspection-jobs/${jobId}`, payload);
  return response.data.data;
}

export async function deleteDiskInspectionJob(jobId: number): Promise<void> {
  await apiClient.delete<ApiResponse<null>>(`/api/disk-monitor/inspection-jobs/${jobId}`);
}

export async function runDiskInspectionJob(jobId: number): Promise<DiskInspectionRun> {
  const response = await apiClient.post<ApiResponse<DiskInspectionRun>>(`/api/disk-monitor/inspection-jobs/${jobId}/run`);
  return response.data.data;
}

export async function fetchDiskInspectionRuns(params: DiskInspectionRunQuery): Promise<DiskInspectionRunPage> {
  const response = await apiClient.get<ApiResponse<DiskInspectionRunPage>>("/api/disk-monitor/inspection-runs", { params });
  return response.data.data;
}

export async function fetchDiskInspectionRun(runId: number): Promise<DiskInspectionRun> {
  const response = await apiClient.get<ApiResponse<DiskInspectionRun>>(`/api/disk-monitor/inspection-runs/${runId}`);
  return response.data.data;
}
